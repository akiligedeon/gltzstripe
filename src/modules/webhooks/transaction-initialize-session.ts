import {
  getEnvironmentFromKey,
  getStripeExternalUrlForIntentId,
  initializeStripePaymentIntent,
  stripePaymentIntentToTransactionResult,
  transactionSessionInitializeEventToStripeCreate,
} from "../stripe/stripe-api";
import { getSaleorAmountFromStripeAmount } from "../stripe/currencies";
import { getWebhookPaymentAppConfigurator } from "../payment-app-configuration/payment-app-configuration-factory";
import { paymentAppFullyConfiguredEntrySchema } from "../payment-app-configuration/config-entry";
import { obfuscateConfig, obfuscateValue } from "../app-configuration/utils";
import { getConfigurationForChannel } from "../payment-app-configuration/payment-app-configuration";
import { type TransactionInitializeSessionResponse } from "@/schemas/TransactionInitializeSession/TransactionInitializeSessionResponse.mjs";
import { type TransactionInitializeSessionEventFragment } from "generated/graphql";
import { invariant } from "@/lib/invariant";
import { createLogger } from "@/lib/logger";

export const TransactionInitializeSessionWebhookHandler = async (
  event: TransactionInitializeSessionEventFragment,
  saleorApiUrl: string,
): Promise<TransactionInitializeSessionResponse> => {
  const logger = createLogger(
    { saleorApiUrl },
    { msgPrefix: "[TransactionInitializeSessionWebhookHandler] " },
  );
  logger.debug(
    {
      transaction: event.transaction,
      action: event.action,
      sourceObject: {
        id: event.sourceObject.id,
        channel: event.sourceObject.channel,
        __typename: event.sourceObject.__typename,
      },
      merchantReference: event.merchantReference,
    },
    "Received event",
  );

  const app = event.recipient;
  invariant(app, "Missing event.recipient!");

  const { privateMetadata } = app;
  const configurator = getWebhookPaymentAppConfigurator({ privateMetadata }, saleorApiUrl);
  const appConfig = await configurator.getConfig();

  const stripeConfig = paymentAppFullyConfiguredEntrySchema.parse(
    getConfigurationForChannel(appConfig, event.sourceObject.channel.id),
  );

  logger.info({}, "Processing Transaction Initialize request");

  const paymentIntentCreateParams = transactionSessionInitializeEventToStripeCreate(event);
  logger.debug({
    paymentIntentCreateParams: obfuscateConfig(paymentIntentCreateParams),
    environment: getEnvironmentFromKey(stripeConfig.publishableKey),
  });

  const stripePaymentIntent = await initializeStripePaymentIntent({
    paymentIntentCreateParams,
    secretKey: stripeConfig.secretKey,
  });

  const data = {
    paymentIntent: { client_secret: stripePaymentIntent.client_secret },
    publishableKey: stripeConfig.publishableKey,
  };
  logger.debug(
    {
      paymentIntent: {
        client_secret: data.paymentIntent.client_secret
          ? obfuscateValue(data.paymentIntent.client_secret)
          : "",
      },
      publishableKey: obfuscateValue(data.publishableKey),
    },
    "Transaction Initialize response",
  );

  const result = stripePaymentIntentToTransactionResult(
    event.action.actionType,
    stripePaymentIntent,
  );
  logger.debug(result, "Stripe -> Transaction result");

  const transactionInitializeSessionResponse: TransactionInitializeSessionResponse = {
    data,
    pspReference: stripePaymentIntent.id,
    result,
    amount: stripePaymentIntent.amount
      ? getSaleorAmountFromStripeAmount({
          amount: stripePaymentIntent.amount,
          currency: stripePaymentIntent.currency,
        })
      : 0,
    time: stripePaymentIntent.created
      ? new Date(stripePaymentIntent.created * 1000).toISOString()
      : undefined,
    message: stripePaymentIntent.cancellation_reason || stripePaymentIntent.description || "",
    externalUrl: stripePaymentIntent.id
      ? getStripeExternalUrlForIntentId(stripePaymentIntent.id)
      : undefined,
  };

  return transactionInitializeSessionResponse;
};
