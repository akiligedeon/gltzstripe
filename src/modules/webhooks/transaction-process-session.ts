import { obfuscateConfig, obfuscateValue } from "../app-configuration/utils";
import { paymentAppFullyConfiguredEntrySchema } from "../payment-app-configuration/config-entry";
import { getConfigurationForChannel } from "../payment-app-configuration/payment-app-configuration";
import { getWebhookPaymentAppConfigurator } from "../payment-app-configuration/payment-app-configuration-factory";
import {
  transactionSessionProcessEventToStripeUpdate,
  getEnvironmentFromKey,
  stripePaymentIntentToTransactionResult,
  getStripeExternalUrlForIntentId,
  updateStripePaymentIntent,
} from "../stripe/stripe-api";
import { getSaleorAmountFromStripeAmount } from "../stripe/currencies";
import { type TransactionProcessSessionEventFragment } from "generated/graphql";
import { type TransactionProcessSessionResponse } from "@/schemas/TransactionProcessSession/TransactionProcessSessionResponse.mjs";
import { createLogger } from "@/lib/logger";
import { invariant } from "@/lib/invariant";

export const TransactionProcessSessionWebhookHandler = async (
  event: TransactionProcessSessionEventFragment,
  saleorApiUrl: string,
): Promise<TransactionProcessSessionResponse> => {
  const logger = createLogger({}, { msgPrefix: `[TransactionProcessSessionWebhookHandler] ` });
  logger.debug(
    {
      transaction: event.transaction,
      action: event.action,
      sourceObject: {
        id: event.sourceObject.id,
        channel: event.sourceObject.channel,
        __typename: event.sourceObject.__typename,
      },
      merchantReference: event.merchantReference,
    },
    `Received event`,
  );

  const app = event.recipient;
  invariant(app, `Missing event.recipient!`);

  const { privateMetadata } = app;
  const configurator = getWebhookPaymentAppConfigurator({ privateMetadata }, saleorApiUrl);
  const appConfig = await configurator.getConfig();

  const stripeConfig = paymentAppFullyConfiguredEntrySchema.parse(
    getConfigurationForChannel(appConfig, event.sourceObject.channel.id),
  );

  logger.info({}, `Processing Transaction Initialize request`);

  const paymentIntentUpdateParams = transactionSessionProcessEventToStripeUpdate(event);
  logger.debug({
    paymentIntentUpdateParams: obfuscateConfig(paymentIntentUpdateParams),
    environment: getEnvironmentFromKey(stripeConfig.publishableKey),
  });

  const stripePaymentIntent = await updateStripePaymentIntent({
    intentId: event.transaction.pspReference,
    paymentIntentUpdateParams,
    secretKey: stripeConfig.secretKey,
  });

  const data = {
    paymentIntent: { client_secret: stripePaymentIntent.client_secret },
    publishableKey: stripeConfig.publishableKey,
  };
  logger.debug(
    {
      paymentIntent: {
        client_secret: data.paymentIntent.client_secret
          ? obfuscateValue(data.paymentIntent.client_secret)
          : "",
      },
      publishableKey: obfuscateValue(data.publishableKey),
    },
    `Transaction Process response`,
  );

  const result = stripePaymentIntentToTransactionResult(
    event.action.actionType,
    stripePaymentIntent,
  );
  logger.debug(result, "Stripe -> Transaction result");

  const transactionProcessSessionResponse: TransactionProcessSessionResponse = {
    data,
    pspReference: stripePaymentIntent.id,
    result,
    amount: stripePaymentIntent.amount
      ? getSaleorAmountFromStripeAmount({
          amount: stripePaymentIntent.amount,
          currency: stripePaymentIntent.currency,
        })
      : 0,
    message: stripePaymentIntent.cancellation_reason || stripePaymentIntent.description || "",
    externalUrl: stripePaymentIntent.id
      ? getStripeExternalUrlForIntentId(stripePaymentIntent.id)
      : undefined,
  };
  return transactionProcessSessionResponse;
};
