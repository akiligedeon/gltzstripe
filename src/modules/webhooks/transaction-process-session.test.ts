import { describe, it, expect } from "vitest";

import { TransactionProcessSessionWebhookHandler } from "./transaction-process-session";
import {
  createMockTransactionProcessSessionEvent,
  createMockTransactionInitializeSessionSourceObjectCheckout,
  createMockTransactionInitializeSessionSourceObjectOrder,
  createMockTransactionInitializeSessionEvent,
} from "./__tests__/utils";
import { TransactionInitializeSessionWebhookHandler } from "./transaction-initialize-session";
import { setupRecording } from "@/__tests__/polly";
import { testEnv } from "@/__tests__/test-env.mjs";

import { TransactionFlowStrategyEnum } from "generated/graphql";

describe("TransactionProcessSessionWebhookHandler", () => {
  setupRecording({});

  describe.each([
    {
      name: "Checkout",
      getSourceObject: createMockTransactionInitializeSessionSourceObjectCheckout,
    },
    { name: "Order", getSourceObject: createMockTransactionInitializeSessionSourceObjectOrder },
  ])("$name", ({ getSourceObject }) => {
    it.each([
      {
        title: "should work authorization",
        data: {},
        result: "AUTHORIZATION_ACTION_REQUIRED",
        amount: 99.99 + 123.0,
        actionType: TransactionFlowStrategyEnum.Authorization,
      },
      {
        title: "should work charge",
        data: {},
        result: "CHARGE_ACTION_REQUIRED",
        amount: 99.99 + 123.0,
        actionType: TransactionFlowStrategyEnum.Charge,
      },
    ])("$title", async ({ title, data, result, amount, actionType }) => {
      // Create payment
      const initializeEvent = await createMockTransactionInitializeSessionEvent({
        data,
        sourceObject: getSourceObject(),
        action: {
          actionType,
        },
      });
      const initializeResult = await TransactionInitializeSessionWebhookHandler(
        initializeEvent,
        testEnv.TEST_SALEOR_API_URL,
      );

      // Update payment
      const processEvent = await createMockTransactionProcessSessionEvent({
        data: {},
        sourceObject: getSourceObject(),
        action: {
          amount: amount + 100,
          actionType,
        },
        transaction: {
          pspReference: initializeResult.pspReference,
        },
      });
      const processResult = await TransactionProcessSessionWebhookHandler(
        processEvent,
        testEnv.TEST_SALEOR_API_URL,
      );
      expect(processResult.data).toEqual(expect.any(Object));
      expect(processResult.result).toEqual(result);
      expect(processResult.amount).toEqual(amount);
      expect(processResult.data).toMatchSnapshot(
        {
          paymentIntent: {
            client_secret: expect.any(String),
          },
        },
        title,
      );
    });
  });
});
