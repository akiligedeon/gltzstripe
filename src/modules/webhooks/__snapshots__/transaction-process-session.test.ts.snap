// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TransactionProcessSessionWebhookHandler > 'Checkout' > 'should work authorization' > should work authorization 1`] = `
{
  "paymentIntent": {
    "client_secret": Any<String>,
  },
  "publishableKey": "pk_test_51LVZwxEosEcNBN5mTKD5afBfOzEF1S1T9tMGyfG4sw6vC6adm8VaKph9EGee1Dk1rlSWz9LgOj4nNNLb2CxJS3HT00x3Dx44oB",
}
`;

exports[`TransactionProcessSessionWebhookHandler > 'Checkout' > 'should work charge' > should work charge 1`] = `
{
  "paymentIntent": {
    "client_secret": Any<String>,
  },
  "publishableKey": "pk_test_51LVZwxEosEcNBN5mTKD5afBfOzEF1S1T9tMGyfG4sw6vC6adm8VaKph9EGee1Dk1rlSWz9LgOj4nNNLb2CxJS3HT00x3Dx44oB",
}
`;

exports[`TransactionProcessSessionWebhookHandler > 'Order' > 'should work authorization' > should work authorization 1`] = `
{
  "paymentIntent": {
    "client_secret": Any<String>,
  },
  "publishableKey": "pk_test_51LVZwxEosEcNBN5mTKD5afBfOzEF1S1T9tMGyfG4sw6vC6adm8VaKph9EGee1Dk1rlSWz9LgOj4nNNLb2CxJS3HT00x3Dx44oB",
}
`;

exports[`TransactionProcessSessionWebhookHandler > 'Order' > 'should work charge' > should work charge 1`] = `
{
  "paymentIntent": {
    "client_secret": Any<String>,
  },
  "publishableKey": "pk_test_51LVZwxEosEcNBN5mTKD5afBfOzEF1S1T9tMGyfG4sw6vC6adm8VaKph9EGee1Dk1rlSWz9LgOj4nNNLb2CxJS3HT00x3Dx44oB",
}
`;
